op.go

```diff
@@ -15,42 +15,231 @@
+		var err error
+		v, err = f.Apply(v)
+		if err != nil {
+			return v, err
+		}
+	}
+	return v, nil
+}
+
 // Dot extract the specific key from the map provided; to extract a nested value, use the Dot Op in conjunction with the
 // Chain Op
-func Dot(key string) OpFunc {
+func Dot(key string, chainFun ...Op) OpFunc {
 	key = strings.TrimSpace(key)
 	if key == "" {
-		return func(in []byte) ([]byte, error) { return in, nil }
+		return func(in reflect.Value) (reflect.Value, error) { return callChain(in, chainFun...) }
+	}
+
+	return func(in reflect.Value) (reflect.Value, error) {
+		for in.Kind() == reflect.Interface || in.Kind() == reflect.Ptr {
+			in = in.Elem()
+		}
+		switch in.Kind() {
+		case reflect.Map:
+			var err error
+			mapVal := in.MapIndex(reflect.ValueOf(key))
+			newMapVal := reflect.New(in.MapIndex(reflect.ValueOf(key)).Type())
+			newMapVal.Elem().Set(mapVal)
+			mapVal = newMapVal.Elem()
+			mapVal, err = callChain(mapVal, chainFun...)
+			if err != nil {
+				return reflect.Value{}, err
+			}
+			in.SetMapIndex(reflect.ValueOf(key), mapVal)
+			return mapVal, nil
+		case reflect.Struct:
+			var r reflect.Value
+
+			if idx, ok := getJSONTag(in, key); ok {
+				r = in.Field(idx)
+			} else {
+				r = in.FieldByName(strings.Title(key))
+			}
+			if r.Kind() == reflect.Invalid {
+				break
+			}
+			return callChain(r, chainFun...)
+		case reflect.Slice:
+			return reflect.Value{}, errors.New("cannot access name field on slice")
+		}
+		return reflect.Value{}, errors.New("key not found")
+	}
+}
+
+// Addition adds the val parameter to the provided interface{} (map/slice/struct)
+func Addition(val interface{}, chainFun ...Op) OpFunc {
+	valRef := reflect.ValueOf(val)
+
+	return func(in reflect.Value) (reflect.Value, error) {
+		// check if in is a pointer/interface
+		// when val is a pointer, set the value of the pointer
+		// when val is a value, set the underlying value of the pointer
+		if (in.Kind() == reflect.Interface || in.Kind() == reflect.Ptr) &&
+			(valRef.Kind() != reflect.Interface && valRef.Kind() != reflect.Ptr) {
+			in = in.Elem()
+		}
+		if !in.CanSet() {
+			return reflect.Value{}, ErrCannotSet
+		}
+
+		switch in.Kind() {
+		case reflect.Slice:
+			if v, ok := val.(json.RawMessage); ok {
+				slcPtr := reflect.New(in.Type())
+				d := json.NewDecoder(bytes.NewReader(v))
+				d.DisallowUnknownFields()
+				err := d.Decode(slcPtr.Interface())
+				if err != nil {
+					return reflect.Value{}, err
+				}
+				valRef = slcPtr.Elem()
+			}
+
+			if valRef.Kind() != reflect.Slice {
+				return reflect.Value{}, fmt.Errorf("Addition: In is slice but val is: %v", valRef.Kind())
+			}
+			if in.Type() != valRef.Type() {
+				return reflect.Value{}, fmt.Errorf("Addition: cannot add elem from slice of type %v to slice of type %v", valRef.Type(), in.Type())
+			}
+			in.Set(reflect.AppendSlice(in, valRef))
+			return callChain(in, chainFun...)
+		case reflect.Map:
+			if v, ok := val.(json.RawMessage); ok {
```

Comment:
`case reflect.Slice` has similar code. We should implement a `valueFromJSON` method